### PR TITLE
Feature: CI with Docker

### PR DESCRIPTION
**Pull request by sxllwx**
**What this PR does**:

Use docker as an infrastructure provider, docker provides dependent third-party components such as Zookeeper, etcdv3, nacos, and consul. Reduce the complexity of Mock in unit testing.

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:
WIP, it is still being tested and modified, and cannot be merged directly at this time.
**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

**Comment by sxllwx** [comment]
> what

**Comment by zouyx** [comment]
> > what
> 
> test / integrate is a sample developed separately for integration testing and works similar to helloworld, using this project as a test sample in the Dockerfile for integration testing. The Server quits after accepting a Client connection and completes a complete interaction, and the Client quits after making a request to the Server and getting a return。 @zouyx

In my opion. That should be a integration test . So move it to https://github.com/apache/dubbo-samples/tree/master/golang will better.

**Comment by sxllwx** [comment]
> I am not sure what `test/integrate` will use for ?



> > > what
> > 
> > 
> > test / integrate is a sample developed separately for integration testing and works similar to helloworld, using this project as a test sample in the Dockerfile for integration testing. The Server quits after accepting a Client connection and completes a complete interaction, and the Client quits after making a request to the Server and getting a return。 @zouyx
> 
> In my opion. That should be a integration test . So move it to https://github.com/apache/dubbo-samples/tree/master/golang will better.
> 
> What do you think? @hxmhlt @fangyincheng @flycash



> > > what
> > 
> > 
> > test / integrate is a sample developed separately for integration testing and works similar to helloworld, using this project as a test sample in the Dockerfile for integration testing. The Server quits after accepting a Client connection and completes a complete interaction, and the Client quits after making a request to the Server and getting a return。 @zouyx
> 
> In my opion. That should be a integration test . So move it to https://github.com/apache/dubbo-samples/tree/master/golang will better.
> 
> What do you think? @hxmhlt @fangyincheng @flycash



> > > what
> > 
> > 
> > test / integrate is a sample developed separately for integration testing and works similar to helloworld, using this project as a test sample in the Dockerfile for integration testing. The Server quits after accepting a Client connection and completes a complete interaction, and the Client quits after making a request to the Server and getting a return。 @zouyx
> 
> In my opion. That should be a integration test . So move it to https://github.com/apache/dubbo-samples/tree/master/golang will better.
> 
> What do you think? @hxmhlt @fangyincheng @flycash

I have considered this solution before, but because cross-code repo collaboration is more difficult to do, and integration testing needs to be based on the code of the Fork repo in the specific PR. This form of CI is more difficult.

File: test/integrate/dubbo/go-server/go.mod [hunk 1]
```diff
@@ -0,0 +1,3 @@
+module github.com/apache/dubbo-go/test/integrate/dubbo/go-server
+
+go 1.14
```
Comment (zouyx) [inline test/integrate/dubbo/go-server/go.mod:3]: in main package go version is 1.13 , should u change it? if not , may cause compile problem.

https://github.com/apache/dubbo-go/blob/master/go.mod

File: test/integrate/dubbo/go-client/go.mod [hunk 2]
```diff
@@ -0,0 +1,3 @@
+module github.com/apache/dubbo-go/test/integrate/dubbo/go-client
+
+go 1.14
```
Comment (zouyx) [inline test/integrate/dubbo/go-client/go.mod:3]: in main package go version is 1.13 , should u change it? if not , may cause compile problem.

https://github.com/apache/dubbo-go/blob/master/go.mod

File: test/integrate/dubbo/go-client/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,21 @@
+FROM golang
+
+MAINTAINER scottwangsxll@gmail.com
```
Comment (zouyx) [inline test/integrate/dubbo/go-client/Dockerfile:3]: what is it mean?

File: test/integrate/dubbo/go-client/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,21 @@
+FROM golang
```
Comment (fangyincheng) [inline test/integrate/dubbo/go-client/Dockerfile:1]: is licence necessary?

File: test/integrate/dubbo/go-client/client.go [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+		case <-time.After(time.Minute):
+			panic("provider not start after client already running 1min")
+		}
+	}()
+	user := &User{}
+	err := userProvider.GetUser(context.TODO(), []interface{}{"A001"}, user)
+	if err != nil {
+		panic(err)
+	}
+	println("response result: %v\n", user)
```
Comment (zouyx) [inline test/integrate/dubbo/go-client/client.go:69]: should you check result to keep it correct?

File: test/integrate/dubbo/go-server/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,20 @@
+FROM golang
```
Comment (zouyx) [inline test/integrate/dubbo/go-server/Dockerfile:1]: Lisence?

File: test/integrate/dubbo/go-server/user.go [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+	Name string
+	Age  int32
+	Time time.Time
+}
+
+type UserProvider struct {
+}
+
+func (u *UserProvider) GetUser(ctx context.Context, req []interface{}) (*User, error) {
+	println("req:%#v", req)
```
Comment (zouyx) [inline test/integrate/dubbo/go-server/user.go:48]: Should you check params to keep it correct?

File: test/integrate/dubbo/go-server/Dockerfile (hunk 6 above)
Comment (sxllwx) [reply test/integrate/dubbo/go-server/Dockerfile]: Okay, I have modified it in the latest commit.

File: test/integrate/dubbo/go-client/Dockerfile (hunk 4 above)
Comment (sxllwx) [reply test/integrate/dubbo/go-client/Dockerfile]: Okay, I have modified it in the latest commit.

File: test/integrate/dubbo/go-client/Dockerfile (hunk 3 above)
Comment (sxllwx) [reply test/integrate/dubbo/go-client/Dockerfile]: Okay, I have modified it in the latest commit.

File: test/integrate/dubbo/go-client/go.mod (hunk 2 above)
Comment (sxllwx) [reply test/integrate/dubbo/go-client/go.mod]: Okay, I have modified it in the latest commit.

File: test/integrate/dubbo/go-server/go.mod (hunk 1 above)
Comment (sxllwx) [reply test/integrate/dubbo/go-server/go.mod]: Okay, I have modified it in the latest commit.

File: test/integrate/dubbo/go-client/client.go (hunk 5 above)
Comment (sxllwx) [reply test/integrate/dubbo/go-client/client.go:69]: I don't think the code here should be hard-coded. result depend provider impl.

File: test/integrate/dubbo/go-server/user.go (hunk 7 above)
Comment (sxllwx) [reply test/integrate/dubbo/go-server/user.go:48]: I don't think the code here should be hard-coded. result depend consumer impl.

File: .travis.yml [hunk 8]
```diff
@@ -1,23 +1,47 @@
...
   - go mod vendor && go test ./... -coverprofile=coverage.txt -covermode=atomic
+  # integrate-test
+  - echo 'start integrate-test'
+  # start zookeeper registry insecure listen in [:]:2181
+  - docker run -d --network host zookeeper
+  - ROOTDIR=$(pwd)
+  - cd ./test/integrate/dubbo/go-client && docker build . -t  ci-consumer --build-arg PR_ORIGIN_REPO=${TRAVIS_PULL_REQUEST_SLUG} --build-arg PR_ORIGIN_COMMITID=${TRAVIS_PULL_REQUEST_SHA} && cd $ROOTDIR
+  - cd ./test/integrate/dubbo/go-server && docker build . -t  ci-provider --build-arg PR_ORIGIN_REPO=${TRAVIS_PULL_REQUEST_SLUG} --build-arg PR_ORIGIN_COMMITID=${TRAVIS_PULL_REQUEST_SHA} && cd $ROOTDIR
+  - docker run -d --network host ci-provider
+  - docker run -it --network host ci-consumer
```
Comment (Patrick0308) [inline .travis.yml:36]: `docker run -d --network host ci-provider` run container in background. when `ci-consumer` started, `ci-provider` may be not boot completed.

File: test/integrate/dubbo/go-client/client.go (hunk 5 above)
Comment (Patrick0308) [reply test/integrate/dubbo/go-client/client.go:69]: I agree with @zouyx . I think should make sure the content returned is what we expect.

File: .travis.yml (hunk 8 above)
Comment (sxllwx) [reply .travis.yml:36]: if ci-provider not boot completed, the ci flow will faild, zookeeper container have sample problem. but our registry will retry.

File: test/integrate/dubbo/go-server/user.go (hunk 7 above)
Comment (sxllwx) [reply test/integrate/dubbo/go-server/user.go:48]: Okay, I have modified it in the latest commit.

File: test/integrate/dubbo/go-client/client.yml [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+      tcp_keep_alive: true
+      keep_alive_period: "120s"
+      tcp_r_buf_size: 262144
+      tcp_w_buf_size: 65536
+      pkg_rq_size: 1024
+      pkg_wq_size: 512
+      tcp_read_timeout: "1s"
+      tcp_write_timeout: "5s"
+      wait_timeout: "1s"
+      max_msg_len: 10240
```
[the remark this answers is not in the document]
Comment (sxllwx) [reply test/integrate/dubbo/go-client/client.yml]: Okay, I have modified it in the latest commit.

**Pull request merged**